# Generating databases and other large files

Download GFF3 files, needed for locating the genes
```bash
ncbi-genome-download --assembly-level complete --flat-output bacteria -F gff
```

Since we already have a fasta database (referenceseeker) we can cross with `transfer/outgoing/databases/referenceseeker/` to see which GFFs have an equivalent FASTA.
In the example below we chose 1k random files, but we can use all (currently 21k) files.
```bash
\ls GCF_* | cut -d "." -f 1 | grep -f referenceseeker.names > both_here_and_refseq.names
for i in `shuf both_here_and_refseq.names | head -n 1000`; do cp ${i}* small/; cp /home/nbi_transfer/outgoing/databases/referenceseeker/bacteria-refseq/${i}* small/; done
```

Alternatively one can use `ncbi-genome-download` again for the fasta, or https://github.com/lskatz/Kalamari. I'll
download the kalamari tables and restrict the blast database to them 
([commit 459a5d1644dee4f2a57252f1de7830ab8fd95a64](https://github.com/lskatz/Kalamari/commit/459a5d1644dee4f2a57252f1de7830ab8fd95a64)).
Notice that there are ~300 chromosomes and ~10k plasmids in the kalamari DB.


To map the fasta file to their contents, we create a table `<filename> <header>`:
```bash
for i in ../gff3/referenceseeker/GCF_*; do zcat $i | grep ">" | cut -c 2- | gawk -v var=$i '{print var,$0}' | perl -pe "s/.*?referenceseeker\///;" ; done > refseeker_headercontents.txt
```

Concatenate fasta files represented in kalamari:
```bash
cut -f 2 kalamari.tsv > kalamari_accession.txt # kalamari.tsv has concatenated `choromosone.tsv` and `plasmid.tsv`
for i in `grep -f kalamari_accession.txt refseeker_headercontents.txt | cut -d' ' -f 1 | sort | uniq`; do zcat ../gff3/referenceseeker/$i >> kalamari.fasta; done

# create blast DB 
makeblastdb -dbtype nucl -in kalamari.fasta -title "files from kalamari chrom and plasmids"
```
